File: week01/red-sprout/solution01.py
def findIdx(n, words):
    l = len(words)
    set_ = set()
    last = ""

    for i in range(l):
        if words[i] in set_:
            return i

        if last != "" and last[-1] != words[i][0]:
            return i

        set_.add(words[i])
        last = words[i]

    return -1


def convert(n, idx):
    if idx == -1:
        return [0, 0]

    return [
        idx % n + 1,
        idx // n + 1
    ]


def solution(n, words):
    idx = findIdx(n, words)
    return convert(n, idx)

File: week01/red-sprout/solution01.java
import java.util.*;

class Solution {
    int findIdx(int n, String[] words) {
        int l = words.length;
        Set<String> set = new HashSet<>();
        String last = "";
        for(int i = 0; i < l; ++i) {
            if(set.contains(words[i])) {
                return i;
            }

            if(!last.equals("")
               && last.charAt(last.length() - 1) != words[i].charAt(0)) {
                return i;
            }

            set.add(words[i]);
            last = words[i];
        }
        return -1;
    }

    int[] convert(int n, int idx) {
        if(idx == -1) {
            return new int[] {0, 0};
        }

        return new int[] {
            idx % n + 1,
            idx / n + 1
        };
    }

    public int[] solution(int n, String[] words) {
        int idx = findIdx(n, words);
        return convert(n, idx);
    }
}
